src/lib/Imp.hs

```diff
@@ -269,178 +307,311 @@ toImpHof env (maybeDest, hof) = do
+makeDest :: AllocInfo -> Type -> Embed ([(Binder, Atom)], Dest)
+makeDest allocInfo ty = do
+  (dest, ptrs) <- flip runCatT mempty $ flip runReaderT env $ makeDestRec ty
+  ptrs' <- forM (envPairs ptrs) $ \(v, (ptrTy, numel)) -> do
+    numel' <- emitBlock numel
+    return (Bind (v:>ptrTy), numel')
+  return (ptrs', dest)
+  where env = DestEnv allocInfo mempty mempty
+
+makeDestRec :: Type -> DestM Dest
+makeDestRec ty = case ty of
+  TabTy b bodyTy -> do
+    depVars <- asks destDepVars
+    if not $ null $ freeVars b `envIntersect` depVars
+      then do
+        (dest, ptrs) <- local (\env -> env { enclosingIdxs = mempty
+                                           , destDepVars   = mempty }) $ scoped $
+                          makeDestRec ty
+        makeBoxes (envPairs ptrs) dest
+      else do
+        lam <- buildLam (Bind ("i":> binderAnn b)) TabArrow $ \(Var i) -> do
+          bodyTy' <- substEmbed (b@>Var i) bodyTy
+          withEnclosingIdxs (Bind i) $ makeDestRec bodyTy'
+        return $ Con $ TabRef lam
+  TypeCon def params -> do
+    let dcs = applyDataDefParams def params
+    case dcs of
+      [] -> error "Void type not allowed"
+      [DataConDef _ bs] -> do
+        dests <- makeDataConDest bs
+        return $ DataConRef def params dests
+      _ -> do
+        tag <- rec TagRepTy
+        let dcs' = applyDataDefParams def params
+        contents <- forM dcs' $ \(DataConDef _ bs) -> forM (toList bs) (rec . binderType)
+        return $ Con $ ConRef $ SumAsProd ty tag contents
+  RecordTy (NoExt types) -> (Con . RecordRef) <$> forM types rec
+  VariantTy (NoExt types) -> do
+    tag <- rec TagRepTy
+    contents <- forM (toList types) rec
+    return $ Con $ ConRef $ SumAsProd ty tag $ map (\x->[x]) contents
+  TC con -> case con of
+    BaseType b -> do
+      ptr <- makeBaseTypePtr b
+      return $ Con $ BaseTypeRef ptr
+    PairType a b -> (Con . ConRef) <$> (PairCon <$> rec a <*> rec b)
+    UnitType     -> (Con . ConRef) <$> return UnitCon
+    CharType     -> (Con . ConRef . CharCon) <$> rec (BaseTy $ Scalar Int8Type)
+    IntRange     l h -> (Con . ConRef . IntRangeVal     l h) <$> rec IdxRepTy
+    IndexRange t l h -> (Con . ConRef . IndexRangeVal t l h) <$> rec IdxRepTy
+    _ -> error $ "not implemented: " ++ pprint con
+  _ -> error $ "not implemented: " ++ pprint ty
+  where rec = makeDestRec
+
+makeBoxes :: [(Name, (Type, Block))] -> Dest -> DestM Dest
+makeBoxes [] dest = return dest
+makeBoxes ((v, (ptrTy, numel)):rest) dest = do
+  ptrPtr <- makeBaseTypePtr $ fromScalarType ptrTy
+  dest' <- makeBoxes rest dest
+  return $ BoxedRef (Bind (v:>ptrTy)) ptrPtr numel dest'
+
+makeBaseTypePtr :: BaseType -> DestM Atom
+makeBaseTypePtr ty = do
+  DestEnv allocInfo idxs _ <- ask
+  numel <- buildScoped $ elemCountE idxs
+  ptrScope <- fmap (const (UnitTy, UnknownBinder)) <$> look
+  scope <- getScope
+  -- We use a different namespace because these will be hoisted to the top
+  -- where they could cast shadows
```

Comment:
FWIW the scoping rules in Imp programs are global, not lexical, so it would be ok to just use `GenName`s here I think.